### DISCUSSION TITLE
Link solidity source code to deployed contract in forked network

### DISCUSSION THREAD

**Discussion by zackwang** [NomicFoundation/hardhat#3656] [Troubleshooting] [answered]
I am debugging my smart contract using local forked mainnet. My contract calls an underlying smart contract which is not deployed by me. When the underlying smart contract fails and throws an exception, the stacktrace only shows `<UnrecognizedContract>.<unknown> (0x contract address)`. I have the source code of the underlying contract and I would like to link it to the deployed binaries so that I know where exactly the exception is thrown. Is there a way to achieve this in hardhat?

**Comment by fvictorio** [comment, marked as answer]
The easiest way to do this is to just add the source code to your project. Make sure that the right compiler options are used; if they don't match the ones used by your contract, you can use [compiler overrides](https://hardhat.org/hardhat-runner/docs/advanced/multiple-solidity-versions).

**Reply by zackwang** [reply]
Thanks for answering! Just curious how does hardhat matches the binaries with source code in the case of forked netowrk?

**Reply by fvictorio** [reply]
Both when forking and when not forking, the basic idea is the same: the bytecode of the contract is analyzed and matched with the bytecodes produced by compiling the project. If there's a match, the source code that generated that bytecode is used.
